_posts/2023-12-20-standard-libraries.markdown

```diff
@@ -0,0 +1,407 @@
+---
+layout: post
+title:  "Dafny Standard Libraries"
+date:   2023-11-20 18:00:00 +0100
+author: Robin Salkeld
+---
+
+I am an engineer and I love Dafny. 
+
+TODO: preamble
+* Intro to Dafny, basic points to back up design later.
+* Intro to adventofcode.com, nice way to close out the year etc.
+* Dafny not a great match for a coding competition where speed is paramount,
+  but this is a great example of how the standard libraries make common tasks so much easier.
+* Other blog posts by my esteemed colleagues (link others) show how to prove interesting properties,
+  but this one is focussing more on how to get things done and support those proof efforts etc.
+
+The requirements for the first puzzle boil down to:
+
+> The newly-improved calibration document consists of lines of text; each line originally contained a specific **calibration value** that the Elves now need to recover. On each line, the calibration value can be found by combining the **first digit** and the **last digit** (in that order) to form a single **two-digit number**.
+>
+> ...
+>
+> Consider your entire calibration document. What is the sum of all of the calibration values?
+
+No problem, we can handle this.
+
+## Getting started
+
+To ensure full backwards compatibility for existing Dafny projects,
+Dafny 4.4 doesn't make the standard libraries available by default:
+they are enabled by a `--standard-libraries` option to the `dafny` CLI.
+But we can do better than that:
+we can use a Dafny project file to indicate we want to use the standard libraries,
+which will let the Dafny IDE understand that dependency.
+
+TODO: set up VS Code extension
+
+Create a new directory with an empty `solution.dfy` file and a `dfyconfig.toml` file with these contents:
+
+```toml
+[options]
+standard-libraries = true
+```
+
+Now for the actual code. Let's start from the bottom-up:
+we're clearly going to need a function to extract the calibration value as a number from a single line of text.
+
+```dafny
+function CalibrationValue(line: string): nat {
+  // ...
+}
+```
+
+How do we find the first and last digits?
+
+## Sequences and Wrappers
+
+Having a look at the [top-level index of Dafny standard libraries](https://github.com/dafny-lang/dafny/tree/master/Source/DafnyStandardLibraries),
+we notice there's a `Std.Collections` library with [a submodule for `Seq`](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/src/Std/Collections/Seq.dfy), and sure enough, there are functions like `IndexOf` and `LastIndexOf`. Perfect!
+
+Or perhaps you were distracted by the `Std.Strings` module, where you might expect these utilities to live.
+But in Dafny, strings are really nothing more than sequences of characters:
+the type `string` is just an alias for `seq<char>`.
+This means many typical string operations are provided as general sequence operations instead.
```

Comment:
That sounds like a design issue. Could we have a module Strings refines Seq so that we not only provide everything Seq provides, but it's discoverable and extensible? I see how methods such as ToUpperCase could be in such a module.

If you agree with this, you can just resolve this comment by cutting an appropriate issue.

Reply:
I agree there's room for improvement here in some way, and I cut this issue with several thoughts that came up while writing this: https://github.com/dafny-lang/dafny/issues/4878